Tenants configurable from the app (the app becoming a default feature).

jtkech opened the pull request:
Maybe fixes #1755 see related comments in this issue. Maybe fixes #1396.

- To be able to configure tenants at the host level, here we have 2 helpers, `ConfigureTenantServices()` and `ConfigureTenant()` with a delegate parameter, which creates and registers a transient module startup at the host level but which is created and executed in tenant contexts.

- `ConfigureTenantServices()` accepts up to 5 dependencies which are injected when the startup instance is created in a tenant context, as aspnet do for some of its helpers.

- So, it's like the application becomes a module with its startup(s) but right now without any feature. But **this allows us to add tenant services and configure tenant pipelines from the host**.

- **This also allows us to replace some modules** which only has a startup. See examples in the cms `ServiceExtensions.cs`. E.g with `OC.Mvc` that could be wrapped e.g in a `AddTenantMvcModules()`. 

            services.AddModules()
                .WithDefaultFeatures(...)
                ...
                ...

                // Replace "OrchardCore.Mvc"
                // Could be wrapped in 'AddTenantMvcModules()'.
                .ConfigureTenantServices<IServiceProvider>((collection, sp) =>
                {
                    collection.AddMvcModules(sp);
                })

                .ConfigureTenant((app, routes, sp) =>
                 {
                     app.UseStaticFilesModules();
                 })

- Then, trying to replace `OC.Commons` i couldn't call `AddTheming()` from the host because it defines shapes that need to be tied to a feature. So, here i continued to **make the application a more complete module which now has one feature which is automatically registered as a default feature**.

- Then i could call `AddTheming()` from the host and all others registration which was done by the `OC.Commons` module. Note: we can't replace `OC.Commons` completely because it also defines a filter, a controller and a view used when we are using the saas recipe.

- So, if it is good to replace some modules, which ones? Some candidates are `".Antiforgery", ".DataProtection" ".Mvc", ".Commons"`. See how they could be replaced in the cms `ServiceExtensions.cs`.

- If we do this, i will do some wrappers helpers, e.g `AddTenantDataProtection()`, and we would need to remove the related dependencies in the existings modules manifests.

**Todo:** See if doing so we could define at the app level some controllers, views, razor pages ...

**Update**: Just tried, after a few updates (not committed), i could write a simple controller view at the app level which works for any tenant (with its current theme). But i need to do more updates because here the app module has an empty subpath, e.g it fails if i update the view at runtime. Hmm, maybe in another PR.

---

- Here we get rid of `ModularServiceCollection` and `ModularApplicationBuilder` to always use regular `IServiceCollection` and `IApplicationBuilder` so that it's easier to use standard / custom DI helpers.

    E.g, in place of having this pattern in `AddOrchardCms()`

        public static IServiceCollection AddOrchardCms(this IServiceCollection services)
        {
            ...
            services.AddModules(modules => 
            {
                modules.WithDefaultFeatures(...);
            });

    We can do

            services.AddModules()
                .WithDefaultFeatures(...)

    And, if people want to add another default feature to their app, they could just use.

            services.AddOrchardCms();
                .WithDefaultFeatures("OtherDefaultFeature")

                // And custom tenant configuration could be added here.
                .ConfigureTenantServices<Dep1, Dep2 ...>((collection, dep1, dep2 ...) =>
                {
                    // some stuff.
                })
                .ConfigureTenant((app, routes, sp) =>
                 {
                    // some stuff.
                 })

---

- `ConfigureTenantServices()` and `ConfigureTenant()` always registers a startup with a transient lifetime. When instances have no injected dependencies, maybe better to register a singleton as aspnet do.

- Here i still use a transient lifetime (that's ok) because when multiple singletons of the same type are registered by the host, there is an issue when cloning them in our child container, could be fixed by #1644.

Jetski5822 commented:
I’m 100% sure I had this idea and Sébastien poo pooed it.

Looking great though!

jtkech commented:
@Jetski5822 :) thanks.

jtkech commented:
@sfmskywalker thanks.

Just wanted to test the possibilities i saw, not even sure myself if all is useful. What is your scenario where you want to do tenant things from the host? 

Right now i search a better way to force the app feature to be registered and which is compatible with all scenarios, because shell features can also be specified from an hard coded list or a configuration file.

I will let know where i am.

sfmskywalker commented:
@jtkech The project I'm working on is a set of microservices built with ASP.NET Core. Each microservice uses Orchard Core's multi-tenancy and modularity features. *All* services expose a web API, which delegate the actual work to some abstraction. The key here is that the actual implementation being injected into a controller depends on which feature is enabled.

Right now, the Web API controllers are implemented in a module, but I would prefer to have these controllers simply be part of the web host application. Only the services that each controller depends on should be provided by modules.

That's my use case :)

jtkech commented:
@sfmskywalker okay cool.

So, i already tested that a controller defined at the host level works here. Then, because here the application is seen as a regular feature, i think that we can decorate the controller (even it is defined at the host level) with a feature attribute and / or a require attribute so that the controller will depend on some features (and the services they register).

But i didn't test this scenario, i will do it as i have time. Hmm, i removed recently the testing code related to view controller defined at the host level ;)

jtkech commented:
@sfmskywalker just tried, it works.

So, i defined a contoller view at the host level which works. Then i decorated it with the feature attribute so that it is served only if the related feature is enabled. Then i did the same test but by using the `[RequireFeatures()]` attribute where you can specify multiple features.

jtkech commented:
@sfmskywalker just for infos.

So the application is automatically registered as a module with one feature that you can't disable. But, after a very little adaptation, i could use our existing manifest feature attribute to register additional features defined at the application level and that you can enable / disable.

As an example, here a controller defined at the application level and which is tied to a feature as described in my previous comment. But here, the feature is also defined at the application level.

    using Microsoft.AspNetCore.Mvc;
    using OrchardCore.Modules.Manifest;

    [assembly: Feature(
        Id = "AppFeature1",
        Description = "Application Feature 1.",
        Category = "Application"
    )]

    namespace OrchardCore.Cms.Web.Controllers
    {
        [Modules.Feature("AppFeature1")]
        public class AppController : Controller
        {
            [Route("About")]
            public IActionResult About()
            {
                return Content("About content");
            }
        }
    }

Then adding this app feature in the blog recipe allows to enable it on the blog setup.

    "steps": [
      {
        "name": "feature",
        "disable": [],
        "enable": [
          // SaaS
          "AppFeature1",
          "OrchardCore.HomeRoute",
          "OrchardCore.Admin",

So, the main application feature can't be disabled (it may add core services) but additional application features can be enabled / disabled.

![appfeatures](https://user-images.githubusercontent.com/8586360/40260960-130ba8c8-5afe-11e8-90e1-9209891f93b0.png)

sebastienros commented:
So an application can also add extra features, that is awesome!

jtkech commented:
Thanks.

Yes, and have dependencies (not the main one which can't be disabled) and so on. I didn't think about all possibilities when i started on this, it never ends ;)

Because feature attributes works, this let me think that maybe we could also use the `Module` attribute. So, maybe we could have the same result without some changes i did here to automatically register the application as a module. But this would force the user to write e.g a manifest file, this in place of just using some helpers.

sfmskywalker commented:
Groovy stuff. Can't wait for this to be merged-in.

sebastienros commented:
Shapes too then I assume in the application? Why I like it is because we can have an application with all the common logic and views, plus some modules that are reusable. To do customization you don't need to create another module, just the host app.

paynecrl97 commented:
Could we use these feature to swap out the default setup module? e.g. have a tenant use a custom module as opposed `Orchard.Setup` for the setup process?

And in the same vain- define a list of custom modules/features that are enabled at all times (even when the tenant is uninitialized)

jtkech commented:
For the 1st point normally we already can specify the minimum set of features needed to bootstrap a custom setup. See in `OC.Application.Cms.Targets/ServiceExtensions.cs`, we have.

            services.AddModules(modules => 
            {
                modules.WithDefaultFeatures(
                    "OrchardCore.Antiforgery", "OrchardCore.Mvc", "OrchardCore.Settings",
                    "OrchardCore.Setup", "OrchardCore.Recipes", "OrchardCore.Commons");
            });

Here, with the replaced modules (maybe temporarily) and new helpers it becomes

                .AddModules()
                .WithDefaultFeatures(
                    "OrchardCore.Settings", "OrchardCore.Setup", "OrchardCore.Recipes")

                // This also works
                .WithDefaultFeatures("OrchardCore.Setup", "OrchardCore.Recipes")

Because you only need the features to boostrap the setup. And then, by adding to `OC.Setup` a dependency on `OC.Recipes`, this was also working.

                .WithDefaultFeatures("OrchardCore.Setup")

It works because of the dependencies so the blueprint is ok, but the temporary descriptor created to boostrap the setup will not contains all features as they were enabled. So i think it's better to specify in all cases all needed features to have a synchronized descriptor when boostrapping the setup

                .WithDefaultFeatures("MySetup", "OtherNeededFeatureToRunMySetup", ...)

---

For the second point i need to think about it. When you say `even when the tenant is uninitialized` do you mean when it is disabled and then it will be still accessible through the features which are always enabled?

jtkech commented:
@sebastienros 

> Shapes too then I assume in the application?

I was thinking so, just tried with `IShapeTableProvider` and `IShapeAttributeProvider`, it works.

So, i copied the shapes provider from the demo module to the application. Then in the app startup i did.

    services.AddOrchardCms()
        .ConfigureTenantServices(collection =>
        {
            collection.AddScoped<IShapeTableProvider, DemoShapeProvider>();
            collection.AddShapeAttributes<DemoShapeProvider>();
        });

Then in a blog theme liquid view i added

    {% shape "Foo" %}
    <h3>Baz application attribute shape</h3>
    {% assign baz = "Baz" | shape_new: text: "Text repeated ", count: 3 %}
    <h3>{{ baz | shape_render }}</h3>

Here the result.

![applicationshapes](https://user-images.githubusercontent.com/8586360/40344034-30a543a2-5d93-11e8-8f59-2f953f5da634.png)

jtkech commented:
For infos, i could do a setup by selecting an application recipe (copied from the blog theme). Then i could use a `IRecipeStepHandler` defined by the application. I added 2 host helpers to add / remove features from any selected recipe on setup, `AddRecipeFeaturesStep()` and `RemoveRecipeFeaturesStep()`.

    .AddRecipeFeatures("OrchardCore.Lucene.Worker")
    .RemoveRecipeFeatures("OrchardCore.Lucene")

sfmskywalker commented:
Why is it called “recipe feature” rather than just “feature”?

jtkech commented:
@sfmskywalker yes you're right.

Maybe just because there are already `WithDefaultFeatures()` used to boostrap a setup and `WithFeatures()` used to define a single tenant application. So to differentiate it and because it relies on recipes.

Anyway, @sebastienros said to me that at this level we should not rely on recipes, so i will remove these helpers. Hmm, on another side these helpers were coming from `OC.Recpes.Implementations`.

But it is planned to add the same kind of helpers but without relying on recipes, and maybe with some logic to filter the tenants on which it could be applyed.

jtkech commented:
@sebastienros where i am.

- Finally i began with an `OrchardCoreBuilder` and an `AddOrchardCore()` trying to add the minimum services. We can't extract `AddLocalization()` because it is used e.g by extension manager, we can't extract e.g `AddWebEncoders()`. I could just extract `AddLogging()` and `AddOptions()`, so i kept them.

- Then, for a more complete `AddOrchard()` which would contain e.g mvc, theming and so on, we would need to reference the related projects, so it can't be done e.g in `OC.Modules`, `OC.Mvc.Core` ... But i think that this helper is the one in `OC.Application.Cms.Targets.ServiceExtensions`. The already existing `AddOrchardCms()` which becomes.

      public static IServiceCollection AddOrchardCms(this IServiceCollection services)
      {
          return services
              .AddCommands()
              // more host level services
              // .Add???()
              .AddOrchardCore(builder => builder // this is an 'OrchardCoreBuilder'.

                  // builder helpers which add host and tenant level services & configuration.
                  .WithDefaultFeatures("OrchardCore.Setup")

                  .AddMvc()
                  .AddAntiForgery()
                  .AddAuthentication()
                  .AddDataProtection()

                  .AddDataAccess()
                  .AddDataStorage()
                  .AddBackgroundTasks()
                  .AddDeferredTasks()

                  .AddTheming()
                  .AddLiquidViews()
                  .AddResourceManagement()
                  .AddCaching()

                  // another way to add more host services
                  // builder.Services.Add???();
                    
                  // more tenant level services
                  .ConfigureTenantServices(collection =>
                  {
                      //collection.Add???();
                  }, 10) // optional order
                  // more tenant level configuration
                  .ConfigureTenant((app, routes, sp) =>
                  {
                      //app.Use???;
                  }), 10);
      }

**Todo**

- I will create a `Startup` helper on which we could use e.g `ConfigureServices`. Right now we still have `ConfigureTenantServices()` and `ConfigureTenant()`. Hmm, maybe good to keep the 2 types if somewhere we want to add services and configuration but with not the same `Order`.

- About limiting the startups, right now the simplest solution is to group some helpers into one, e.g all security helpers in a `AddSecurity()`. There are not so much startups (knowing there is less modules), but if it is worth to limit them i will think about it.

- Create `AddEnabledFeatures()` with the ability to add some tenant logic to select the features to enable. Hmm, maybe here we could use the `ShellOptions` on which we could add an `EnabledFeatures` property that we could update, i will see.

sebastienros commented:
AddOrchard should not have Mvc, for instance we could just want modules and tenants with middlewares, or with nancy. This should be added by AddOrchardCms.

jtkech commented:
Okay for `AddOrchard()` without Mvc. And in the above example of `AddOrchardCms()` it is added through the delegate of `AddOrcharCore()` which doesn't add Mvc itself.

But, as i saw, there would be no significant difference between an `AddOrchard()`, which i didn't create yet, and the `AddOrchardCore()` i just created. E.g where we could define `AddOrchard()` e.g with theming (as you said to me), here we would need to reference `OC.DisplayManagement`.

jtkech commented:
And just saw that `OC.DisplayManagement` references `OC.Mvc.Core`.

jtkech commented:
About the startups one idea would be to enlist all `ConfigureTenantServices()` and `ConfigureTenant()` so that, if they use the same order (e.g the default), the application will have only one pseudo startup with this order. If 2 different orders are used, 2 startups will be created and registered, and so on. I will try this.

Then i will create the ability to write

    builder.Startup
        .ConfigureServices(collection => ..., optionalOrder)
        .Configure((app, routes, sp) => ..., optionalOrder)

As a side note an application can define a module startup as a module do but this class would need to belong to the app and this can't be defined automatically through helpers coming from any libraries.

jtkech commented:
@sebastienros you were right, much more simple to have only one version of `ConfigureTenantServices()` which inject an `IServiceProvider`, this in place of 5 more versions where you can specify up to 5 dependency types. I did the change.

jtkech commented:
@sebastienros as a summary of what i did.

- Group pseudos module startups (registered by the app) of the same `Order` in only one.

- Only one version of pseudo startups `ConfigureTenantServices()` which inject an `IServiceProvider`, this in place of having 5 more versions where you can specify up to 5 dependency types.

- New startup builder which allow to write.

      .AddOrchardCore(builder => builder
          ...
          .Startup
              .ConfigureServices((collection, sp) => { })
              .Configure((app, routes, serviceProvider) => { })


**Todo**.

- Create `AddEnabledFeatures()` with the ability to add some tenant logic to select the features to enable.

- Then the last thing would be to create `AddOrchard()`. So i first created `AddOrchardCore()` with the essential needed services. In the `App.Cms.Targets`, the existing `AddOrchardCms()` uses it and adds through the delegate MVC, theming and so on.

- But now i don't know what would be `AddOrchard()` and where i could define it.

jtkech commented:
So, i added a **tenant builder** so that in `builder.Startup.ConfigureServices()` we also have good intellisense when using `AddFoo()` which could have the same name but here only adds tenant services.

For testing and it works, here i extracted and only used the helpers which only add host or tenant services but not both, so that you can see how we can compose.

        public static IServiceCollection AddOrchardCmsTest(this IServiceCollection services)
        {
            return services

                .AddOrchardCore(builder =>
                {
                    builder
                        .AddCommands()
                        .WithDefaultFeatures("OrchardCore.Setup");

                    builder.Services.AddAntiforgery();
                    builder.Services.AddAuthentication();

                    builder
                        .AddSitesFolder()
                        .AddThemingHost()
                        .AddManifestDefinition("theme")

                        .Startup.ConfigureServices((tenant, sp) =>
                        {
                            tenant
                                .AddMvc(sp)
                                .AddAntiForgery(sp)
                                .AddAuthentication()
                                .AddDataProtection(sp)

                                .AddDataAccess()
                                .AddShellDescriptorStorage()
                                .AddBackgroundTasks()
                                .AddDeferredTasks()
                                .AddTheming()
                                .AddLiquidViews()
                                .AddResourceManagement()
                                .AddCaching();

                            tenant.Services.AddTagHelpers(typeof(BaseShapeTagHelper).Assembly);
                            tenant.Services.AddTagHelpers(typeof(ResourcesTagHelper).Assembly);
                        })

                        .Configure((tenant, routes, sp) =>
                        {
                            tenant.UseStaticFiles();
                            tenant.ApplicationBuilder.UseAuthentication();
                        });
                });

        }

---

Then here how `AddOrchardCms()` is when only using helpers which add both host and tenant services.

        public static IServiceCollection AddOrchardCms(this IServiceCollection services)
        {
            return services

                .AddOrchardCore(builder => builder
                    .AddCommands()

                    .AddMvc()
                    .AddAntiForgery()
                    .AddAuthentication()
                    .AddDataProtection()
                    
                    .WithDefaultFeatures("OrchardCore.Setup")

                    .AddDataAccess()
                    .AddDataStorage()
                    .AddBackgroundTasks()
                    .AddDeferredTasks()

                    .AddTheming()
                    .AddLiquidViews()
                    .AddResourceManagement()
                    .AddCaching());
        }

jtkech commented:
@sebastienros just for infos

- I would like to remove the following.

      // MVC is already registering IMemoryCache as host singleton. We are registering it again
      // in this module so that there is one instance for each tenant.
      // Important: we can't call AddMemoryCache as it's using the TryAdd pattern and hence would
      // not override any existing instance defined at the host level by MVC
      services.AddSingleton<IMemoryCache, MemoryCache>();

    `IMemoryCache` is registered by MVC through `AddMvcCore()` and when we call it we already are in the **context of a tenant**. I tried it after removing the above and checked that a different tenant still get a different instance of the memory cache. **But finally i kept it** because i thought about an app without MVC but which still want to use our caching. Let me know.

    In fact `IMemoryCache` is registered through `AddRazorViewEngine()` which is called by `AddViewLocalization()`. Then i saw that `AddViewLocalization()` also calls `AddViews()`, so i commented out our own call to `AddViews()`.

- Cool, new conflicts to resolve ;)

Skrypt commented on src/OrchardCore/OrchardCore.Environment.Cache/TenantServicesBuilderExtensions.cs:
```diff
@@ -26,10 +26,7 @@ public static TenantServicesBuilder AddCaching(this TenantServicesBuilder tenant
+            // IMemoryCache is registered at the tenant level so that there is one instance for each tenant.
             services.AddSingleton<IMemoryCache, MemoryCache>();
 
             // MemoryDistributedCache needs to be registered as a singleton as it owns a MemoryCache instance.
```
I was actually thinking about this and here it is already done. 👍

jtkech replied:
@Skrypt 

About `IMemoryCache`, it is registered by an mvc helper which is called when we call our `AddMvcModules()` (renamed here). In this case, `IMemoryCache` is already registered as a per tenant singleton (not an app singleton). But there are multiple scenarios, e.g:

1. An Orchard mvc app which call our helper => nothing to do.

2. A non mvc Orchard app but want to use our tenant level caching => we need to register it as we do and here we could use `AddMemoryCache()`.

3. Same as 2. but the app also want to use the regular `.AddCaching()` but at the app level which here would add an application singleton that would be also shared by all tenants => we need to register it as we do and here we can't use `AddMemoryCache()` with its `.TryAdd()` pattern.

About `MemoryDistributedCache` it is registered e.g by the mvc helper `AddCacheTagHelper()` but we we don't call it. So, you're right it is not registered in the first above scenario and we have to do it. Then, for the same reason as above we can't use `AddDistributedMemoryCache()`.

Note: Yes, we don't use the aspnet `AddMvc()`, we only call `AddMvcCore()` with the essential mvc services because we need to compose then with mvc and custom orchard services. Hmm, this let me think that we need to revisit what the regular `AddMvc()` now do to be sure we don't loose anything useful, e.g we don't call the regular `AddCacheTagHelper()`, `AddRazorPages()`.

Then, as i remember, we also register as a tenant singleton a `DefaultDynamicCache` which is only a wrapper so, as i understand, it adds nothing in term of implementation but it is registered by the `OC.DynamicCache` so that it is available only if the related feature is enabled and then it will use the currently registered `IDistributedCache`.

Skrypt replied:
I think you got this under control 😉

jtkech commented:
Another example using `AddOrchardCore()` without delegate (the same can be done with a delegate).

            var builder = services.AddOrchardCore();

            // you can add host and tenant level services
            // by using 'OrchardCoreBuilder' helpers.
            builder.AddAuthentication();
            builder.AddTheming();
            builder.AddEnabledFeatures("OrchardCore.Localization"); // so for all tenants.

            // OR you can just add host level services and then
            // add tenant level services afterwards (see below).
            builder.Services.AddAuthentication();
            builder.AddThemingHost();

            builder.Startup.ConfigureServices((tenant, sp) =>
            {
                var tenantName = sp.GetRequiredService<ShellSettings>().Name;

                // some tenant aware logic
                if (tenantName == "Alpha")
                {
                    // here, add services only at the tenant level
                    // by using 'TenantServicesBuilder' helpers.
                    tenant.AddAuthentication();
                    tenant.AddTheming();
                    tenant.AddEnabledFeatures("OrchardCore.Localization"); // so only for this tenant.

                    // add other regular services
                    tenant.Services.AddRegularServices();
                }
            });

            // can be called multiple times
            builder = services.AddOrchardCore();

bsamba commented:
I like the new breaking changes, lol. However, I'm not sure why AddRecipes() was not included, because without it, setting up a new tenant does not work--recipes cannot be resolved. Anyway, I added it as a global feature: AddGlobalFeatures("OrchardCore.Recipes"), and everything works.

jtkech commented:
@bsamba thanks for the info.

Just tried to setup a tenant with our solution, it works on my side. In `AddOrchardCms()` we just use.

    .AddSetupFeatures("OrchardCore.Setup")

And it works because now `OC.Setup` has a dependency on `OC.Recipes`. So, e.g if you use a custom setup module you might have to add a dependency on `OC.Recipes`.

bsamba commented:
@jtkech Indeed, `OC.Setup` has the dependency and ours does not. We'll go with your recommendation, many thanks.
